crates/bifrost/src/providers/replicated_loglet/record_cache.rs

```diff
@@ -22,4 +35,63 @@ impl RecordCache {
     pub fn new(_memory_budget_bytes: usize) -> Self {
         Self {}
     }
+
+    /// Append the next batch to the loglet_id cache.
+    ///
+    /// Cache expecting offsets to come in order. If the provided
+    /// `first_offset` would create a gap in the cache the append
+    /// will be rejected with an error.
+    pub(crate) fn append_records(
+        &self,
+        _loglet_id: ReplicatedLogletId,
+        _first_offset: LogletOffset,
+        _records: Arc<[Record]>,
+    ) -> Result<(), RecordCacheError> {
+        todo!()
+    }
+
+    /// Release offset (inclusive) to readers.
+    ///
+    /// All active reader streams should receive all released batches.
+    pub(crate) fn release_records(&self, _loglet_id: ReplicatedLogletId, _offset: LogletOffset) {
+        todo!()
+    }
+
+    /// Read a range of records.
+    ///
+    /// A Gap can only appear as a first item in the returned
+    /// range.
+    ///
+    /// It's up the the caller to full fill the gap by other means
+    /// for example by reading directly from the log servers.
+    pub fn read_range<R>(&self, _range: R) -> Vec<CachedRecords>
+    where
+        R: RangeBounds<LogletOffset>,
+    {
+        todo!()
+    }
+
+    /// Gets a read stream that starts from the given offset.
+    ///
+    /// The first item returned by the stream can be a gap
+    /// if the `from` offset is not available in the cache
+    /// anymore.
+    pub fn get_read_stream(
```

Comment:
I am rethinking this and I think the stream can return multiple gaps for slow readers in case the cache of some records where dropped. This is to prevent the cache from holding old records for too long